feat: dynamic org-level query concurrency limits

zlwaterfield opened the pull request:
## Changes

Added support for per-organization query concurrency limits using the ORGANIZATION_QUERY_CONCURRENCY_LIMIT feature. Updated the rate limiter to check for this feature and use its value if present, with fallback to defaults. Defaults have stayed the same.

Note: this shouldn't change anything when shipped since no customers actually have the `ORGANIZATION_QUERY_CONCURRENCY_LIMIT` set yet.

## How did you test this code?

Added tests

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

zlwaterfield committed:
Support org-level query concurrency feature limit

Added support for per-organization query concurrency limits using the ORGANIZATION_QUERY_CONCURRENCY_LIMIT feature. Updated the rate limiter to check for this feature and use its value if present, with fallback to defaults. Added comprehensive tests for the new behavior and registered the new feature in AvailableFeature.

zlwaterfield committed:
Add Redis caching for org query concurrency limits

Refactored RateLimit to cache organization-specific query concurrency limits in Redis, reducing database lookups. Added a helper method for fetching and caching the limit, and expanded tests to cover cache hits, misses, and fallback scenarios.

anirudhpillai commented on posthog/clickhouse/client/limit.py:
```diff
@@ -155,6 +167,32 @@ def use(self, *args, **kwargs) -> tuple[Optional[str], Optional[str]]:
 
         return running_tasks_key, task_id
 
+    def _get_org_concurrency_limit(self, org_id: int) -> Optional[int]:
```
the implementation of this function is great (especially with the redis caching of the feature limit)
though I'd prefer if we keep this function separate from `RateLimit` class. Maybe the rate limit class can accept a callable for max_concurrency. When the arg is a callable and not an int, then it uses the callable to determine max concurrency?

zlwaterfield replied:
@anirudhpillai decided against the callable because it made it move convoluted that it needed to be, but I did move it out of the rate limit class.

zlwaterfield committed:
Refactor org concurrency limit logic and tests

Moved organization concurrency limit logic to a standalone helper function `get_org_concurrency_limit` and updated the `RateLimit` class to support dynamic max_concurrency via callables. Enhanced test coverage to verify priority order between beta teams, explicit limits, and organization-specific limits.

zlwaterfield committed:
Merge remote-tracking branch 'origin/master' into zach/dynamic-concurrency-from-feature-entitlements

zlwaterfield committed:
Refactor and expand org concurrency limit tests

Replaces TestAppOrgRateLimiterWithQueryConcurrency with TestOrgConcurrencyLimit, focusing on testing the get_org_concurrency_limit helper function. Adds new tests for cache miss/hit, invalid values, and exception handling. Also adds a test for callable max_concurrency in RateLimit. Removes redundant and less focused tests to streamline coverage.

zlwaterfield committed:
Increase org concurrency limit cache to 1 hour

Changed the Redis cache duration for organization concurrency limits from 5 minutes to 1 hour to reduce cache misses and improve performance.

zlwaterfield committed:
Add org concurrency limit to rate limiter

Passes the organization-specific concurrency limit to the app org rate limiter in QueryRunner. This allows for dynamic control of concurrent queries per organization.

zlwaterfield committed:
Remove callable max_concurrency from RateLimit

Refactored RateLimit to only accept static integer values for max_concurrency, removing support for callables. Updated usage in query_runner.py to pass dynamic limits via the 'limit' argument instead. Removed related tests for callable max_concurrency from test_limit.py.

zlwaterfield committed:
Update test_limit.py

zlwaterfield committed:
UUIDT type

zlwaterfield committed:
Merge branch 'master' into zach/dynamic-concurrency-from-feature-entitlements

zlwaterfield committed:
Update limit.py

zlwaterfield committed:
Fix handling of explicit limit override in RateLimit

Ensures that the explicit 'limit' override in RateLimit is properly cast to int and only used if not None. Also updates the test to use a UUID for non-existent orgs, improving test reliability.

zlwaterfield committed:
Refactor max_concurrency fallback logic in RateLimit

Moved the static default assignment of max_concurrency to the top of the method and removed redundant fallback code. This simplifies the logic for determining max_concurrency and ensures the default is always set initially.

anirudhpillai reviewed (approved):
💯

orian commented:
This looks good to me, only thing i would do is to specify whether the limit is general (all queries), app or api. The rate limiter you've put it is applied only for the queries originating from our app (app.posthog.com)

orian commented on posthog/clickhouse/client/limit.py:
```diff
@@ -96,12 +102,20 @@ def use(self, *args, **kwargs) -> tuple[Optional[str], Optional[str]]:
+        # Static default
+        max_concurrency: int = self.max_concurrency
+
+        # Determine max_concurrency with proper priority order
```
this comment is misleading

zlwaterfield replied:
Ok ill remove

orian commented on posthog/clickhouse/client/limit.py:
```diff
@@ -96,12 +102,20 @@ def use(self, *args, **kwargs) -> tuple[Optional[str], Optional[str]]:
         in_beta = kwargs.get("is_api") and (team_id in settings.API_QUERIES_PER_TEAM)
+
         if in_beta:
+            # Beta teams get highest priority
```
they don't get highest priority, they get a per team limit

zlwaterfield committed:
clean up comments

zlwaterfield committed:
Rename org query concurrency limit to app query limit

Refactored code to rename ORGANIZATION_QUERY_CONCURRENCY_LIMIT to ORGANIZATION_APP_QUERY_CONCURRENCY_LIMIT for clarity and consistency. Updated helper function, cache key, feature flag, and related tests and usages to reflect the new naming.
